### PR TITLE
multi: store KeyLocator in OpenChannel, use ECDH instead of DerivePrivKey

### LINKED ISSUE #2659: channeldb: store the shachain key locator in instead of raw producer
Atm, [we store the shachain root directly in `channeldb`](https://github.com/lightningnetwork/lnd/blob/master/channeldb/channel.go#L481). Fortunately, [we derive this using the proper `keychain` key family, and use the serialize private key directly as the root](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/wallet.go#L551). However, since we don' also store the `keychain.KeyLocator` information, during the implementation of the new SCB format, I was forced to brute force query the wallet in order to obtain the private key that's need in order to carry out the DLP protocol. In the current implementation, we consciously limit the search space to 10k or so, to avoid a very long lived search or an inadvertent infinite loop. Instead, we should store _only_ the `KeyLocator` information so we don' t need to perform this search. As a side effect, we'll also get away from storing this private key on disk in plaintext. 

**NOTE:** fixing this issue will involve a database migration. At this point, we've decided to hold off on all database migrations until we have a TLV system in place as it will make the migrations much more streamlined when we need to modify _values_ already stored in the system. 

# Steps To Completion 

- [ ] Using the yet-to-be-written TLV system, add a new field to the `channeldb.OpenChannel` struct that stores the `keychain.KeyLocator` for out `shachain` root. 

- [ ] Deprecate the old value for all new channels going forward 

- [ ] Update the yet-to-be-merged SCB feature to stop using the brute-force method, in favor of directly deriving the key based off of the key locator.

### PR DESCRIPTION
**Pull request by Crypt-iQ**
Closes #2659
Related #3929 

By adding the `RevocationKeyLocator` in the same PR, we get away with not adding a version to SCB as we know that if an `OpenChannel` has a populated `RevocationKeyLocator`, they are using `ECDH` derivation for the sha chain root key.

The question then is if this is better, or more confusing down the line and a version would come with the least amount of headaches.

### CODE REVIEW

**Comment by Crypt-iQ** [comment]
Seems like the windows build is a flake. Outstanding thing is to test this with a node that has the old channels present in a backup

File: tlv/complex.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package tlv
+
+import (
+	"io"
+
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// EKeyLocator is an encoder for keychain.KeyLocator.
+func EKeyLocator(w io.Writer, val interface{}, buf *[8]byte) error {
```
Comment (guggero) [inline tlv/complex.go:10]: These should have unit tests.

File: tlv/complex.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		}
+		v.Family = keychain.KeyFamily(family)
+
+		return DUint32(r, &v.Index, buf, 4)
+	}
+	return NewTypeForDecodingErr(val, "keychain.KeyLocator", l, 8)
+}
+
+// MakeKeyLocRecord creates a Record out of a KeyLocator using the passed
+// Type and the EKeyLocator and DKeyLocator functions. The size will always be
```
Comment (guggero) [inline tlv/complex.go:38]: nit: missing number after "will always be".

File: chanbackup/single.go [hunk 3]
```diff
@@ -120,19 +120,34 @@ type Single struct {
...
+		// This is the legacy scheme.
+		var b bytes.Buffer
+		_ = channel.RevocationProducer.Encode(&b) // Can't return an error.
 
-	// Once we have the root, we'll make a public key from it, such that
-	// the backups plaintext don't carry any private information. When we
-	// go to recover, we'll present this in order to derive the private
-	// key.
-	_, shaChainPoint := btcec.PrivKeyFromBytes(btcec.S256(), b.Bytes())
+		// Once we have the root, we'll make a public key from it, such that
```
Comment (guggero) [inline chanbackup/single.go:138]: nit: comments overflow 80 chars.

File: chanbackup/single.go [hunk 4]
```diff
@@ -120,19 +120,34 @@ type Single struct {
...
+		// the backups plaintext don't carry any private information. When
+		// we go to recover, we'll present this in order to derive the
+		// private key.
+		_, shaChainPoint := btcec.PrivKeyFromBytes(btcec.S256(), b.Bytes())
+
+		shaChainRootDesc = keychain.KeyDescriptor{
+			PubKey: shaChainPoint,
```
Comment (guggero) [inline chanbackup/single.go:145]: So the way to identify if this is a channel with the old or new scheme when looking at a backup file is whether the  `PubKey` field here is set or not?

File: channeldb/channel.go [hunk 5]
```diff
@@ -3122,6 +3132,27 @@ func putChanInfo(chanBucket kvdb.RwBucket, channel *OpenChannel) error {
 		return err
 	}
 
+	// Write the RevocationKeyLocator as the first entry in a tlv stream.
+	keyLocRecord := tlv.MakeKeyLocRecord(
+		keyLocType, &channel.RevocationKeyLocator,
+	)
+
+	tlvStream, err := tlv.NewStream(
```
Comment (guggero) [inline channeldb/channel.go:3140]: nit: can fit on line line.

File: chanrestore.go [hunk 6]
```diff
@@ -48,19 +48,18 @@ type chanDBRestorer struct {
...
 
-	// First, we'll also need to obtain the private key for the shachain
-	// root from the encoded public key.
-	//
-	// TODO(roasbeef): now adds req for hardware signers to impl
-	// shachain...
-	privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
+	// Derive the shachain root via an ECDH operation on the private key
+	// of ShaChainRootDesc and the remote node's public key.
+	ecdh, err := c.secretKeys.ECDH(backup.ShaChainRootDesc, backup.RemoteNodePub)
```
Comment (guggero) [inline chanrestore.go:53]: What if we have to restore a backup for a channel that was written before this change? Won't that create an invalid shachain root? I think we need to support the legacy was here as well and error out if `DerivePrivKey` isn't available because `lnd` runs in "readonly wallet mode".
We'll probably require a node to have no channels when turning on "readonly wallet mode" to make sure they don't run into this. But we do have to support both backup schemes for a long time IMO.

File: lnwallet/wallet.go [hunk 7]
```diff
@@ -805,18 +805,29 @@ func (l *LightningWallet) initOurContribution(reservation *ChannelReservation,
 	if err != nil {
 		return err
 	}
-	revocationRoot, err := l.DerivePrivKey(nextRevocationKeyDesc)
+
+	if nextRevocationKeyDesc.Index == 0 {
```
Comment (guggero) [inline lnwallet/wallet.go:809]: Why is this needed? Should at least have a comment.

**Review by guggero — CHANGES_REQUESTED** [review]
Did a first pass, looks pretty far along. I like the idea of using TLV to not require a DB migration.

My main comment is about supporting the old scheme in SCB recovery, preferably with an integration test backing it up.
And some unit tests here and there would be nice as well.

File: chanrestore.go (hunk 6 above)
Comment (Crypt-iQ) [reply chanrestore.go:53]: We always use `BtcWalletKeyRing` which may change in the future.  But for now both schemes are supported, probably just need an integration test to back it up.

Comment (guggero) [reply chanrestore.go:53]: I don't see how this would still support "old" channels? With that I mean backups for channels that were created before this PR is merged. If we now always do `ECDH`, the derived revocation root will be wrong for "old" channels.

Comment (Crypt-iQ) [reply chanrestore.go:53]: https://github.com/lightningnetwork/lnd/blob/64b8aa8f114142b4b1d1e8e1e751c846d1a7f422/keychain/btcwallet.go#L362

Comment (Crypt-iQ) [reply chanrestore.go:53]: whoops, guess I thought it was handled for some reason.  How do you think this should be handled?

Comment (guggero) [reply chanrestore.go:53]: I think we have to look at the `PubKey` field of the `ShaChainRootDesc`. If it's nil, it's the new scheme, if it's set, it's the old scheme. If it's the old scheme, we just derive the private key. The PR that removes the `DerivePrivKey` in the future will need a way to deal with that (for example erroring out with "restoring old channels on a read-only lnd wallet is not supported") or whatever.

Comment (Crypt-iQ) [reply chanrestore.go:53]: Yeah you're right. For some reason I thought it was handled. We have a problem then because if we want a completely *external* signer, and they want to restore an old channel, it won't be possible as the external signer would need to know the details of `DerivePrivKey`. But I guess we can handle it? I guess if that is the case, we would have to error out like you said.

**Comment by Crypt-iQ** [comment]
How will we deprecate RevocationProducer? Is the motivation to remove the shachain root from being on-disk? And instead be in memory if possible with ecdh derivation done on startup? Might be missing some context. @Roasbeef @guggero

File: lnwallet/wallet.go (hunk 7 above)
Comment (Crypt-iQ) [reply lnwallet/wallet.go:809]: DerivePrivKey in keychain/btcwallet.go special cases Index == 0 implicitly

File: chanrestore.go (hunk 6 above)
Comment (guggero) [reply chanrestore.go:53]: Yes, when using the completely external signer mode in lnd, we'd need to error out everywhere we still use `DerivePrivKey`, even if it's just a fallback. We just need to advise users (or enforce it somehow) that they need to start a new node if they use that mode to avoid this problem.

**Comment by guggero** [comment]
I don't think we want to deprecate the `RevocationProducer`. It should still be in the channel and be written to disk. Its content doesn't change during the lifetime of the channel. We just change the way it is created in the first place.

**Comment by Crypt-iQ** [comment]
> I don't think we want to deprecate the `RevocationProducer`. It should still be in the channel and be written to disk. Its content doesn't change during the lifetime of the channel. We just change the way it is created in the first place.

My main reasoning is that it creates all the revocations and we ideally don't even have to store that on-disk (only the materials needed for revocation) even though it's encrypted. But I think that's a whole other can of worms and may not even be the most sensitive data stored on disk.

File: chanbackup/single.go (hunk 4 above)
Comment (Crypt-iQ) [reply chanbackup/single.go:145]: Correct - and that's why I think a version might be better so as to be explicit.

**Comment by Crypt-iQ** [comment]
I don't think we can do a legacy recovery in the itests as we can no longer create legacy `RevocationProducer` with new channels.

File: chanbackup/single.go (hunk 3 above)
Comment (Crypt-iQ) [reply chanbackup/single.go:138]: I think I fixed this?

File: lnwallet/wallet.go (hunk 7 above)
Comment (Crypt-iQ) [reply lnwallet/wallet.go:836]: Added a comment

**Comment by Crypt-iQ** [comment]
Tested locally by:
- opening channel on 9e75c3eadede1816d4be19a7923cd5d0b44b78a0
- exchange some HTLCs
- shutdown node
- start node on 900d8042d514e1b5aeea60b7dc95d8c8aa3700e6
- initiate DLP (works)
- create a new channel with ECDH scheme
- exchange some HTLCs
- shutdown node
- startup and initiate DLP again (works)

Remaining TODO is to write the tlv tests

**Comment by guggero** [comment]
> I don't think we can do a legacy recovery in the itests as we can no longer create legacy `RevocationProducer` with new channels.

Would it maybe make sense to create a channel backup with an older version and check that into git, together with the seed it was created with? I think making sure we always support old channel backups in the future is probably a worthy test case.

**Comment by Crypt-iQ** [comment]
> > I don't think we can do a legacy recovery in the itests as we can no longer create legacy `RevocationProducer` with new channels.
> 
> Would it maybe make sense to create a channel backup with an older version and check that into git, together with the seed it was created with? I think making sure we always support old channel backups in the future is probably a worthy test case.

I think that's a good idea - we should check in some old backup + seed, but the problem I ran into is that the itests require a full channel restore which requires another node to also have knowledge of this channel to initiate DLP.

**Comment by guggero** [comment]
Right, we'd probably need to check in a `channel.db` for the channel peer as well (and the seed for it).

**Comment by Crypt-iQ** [comment]
> Right, we'd probably need to check in a `channel.db` for the channel peer as well (and the seed for it).

I think this is a good test to have - I'll start on it

File: chanrestore.go [hunk 8]
```diff
@@ -48,19 +48,40 @@ type chanDBRestorer struct {
...
-	privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
-	if err != nil {
-		return nil, fmt.Errorf("unable to derive shachain root key: %v", err)
-	}
-	revRoot, err := chainhash.NewHash(privKey.Serialize())
-	if err != nil {
-		return nil, err
+	// The shachain root that seeds the RevocationProducer for this channel.
+	// It currently has two possible formats.
+	var revRoot *chainhash.Hash
```
Comment (Roasbeef) [inline chanrestore.go:53]: Style nit: can collapse into a single var statement.

File: chanrestore.go [hunk 9]
```diff
@@ -48,19 +48,40 @@ type chanDBRestorer struct {
...
+	} else {
+		// This is the scheme in which the shachain root is derived via an
+		// ECDH operation on the private key of ShaChainRootDesc and the remote
+		// node's public key.
+		ecdh, err := c.secretKeys.ECDH(backup.ShaChainRootDesc, backup.RemoteNodePub)
+		if err != nil {
+			return nil, fmt.Errorf("unable to derive shachain root: %v", err)
+		}
+
+		revRoot, err = chainhash.NewHash(ecdh[:])
```
Comment (Roasbeef) [inline chanrestore.go:79]: We don't actually need to hash this _again_, since the x-coordinate which is the output of the ECDH is hashed before returning the output.

File: chanrestore.go [hunk 10]
```diff
@@ -48,19 +48,40 @@ type chanDBRestorer struct {
...
+
+		revRoot, err = chainhash.NewHash(privKey.Serialize())
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// This is the scheme in which the shachain root is derived via an
+		// ECDH operation on the private key of ShaChainRootDesc and the remote
+		// node's public key.
+		ecdh, err := c.secretKeys.ECDH(backup.ShaChainRootDesc, backup.RemoteNodePub)
```
Comment (Roasbeef) [inline chanrestore.go:74]: Hmm, I had originally envisioned just doing ECDH against one of our _own_ keys. It seems a bit redundant, but this ensure that the remote party isn't able to somehow re-derive _our_ shachain root if they somehow learn of an SCB.

**Review by Roasbeef — CHANGES_REQUESTED** [review]
The set of changes looks good to me, turned out to be rather straightforward as well. My only main comment is possibly changing _which_ key we ECDH with.

File: chanrestore.go (hunk 10 above)
Comment (Crypt-iQ) [reply chanrestore.go:74]: It should ideally be one of our keys.

File: chanrestore.go (hunk 9 above)
Comment (Crypt-iQ) [reply chanrestore.go:79]: Right, it's already a hash

Comment (Crypt-iQ) [reply chanrestore.go:79]: This will give us the wrong seed as implemented.

Comment (Crypt-iQ) [reply chanrestore.go]: This doesn't hash it as far as I can tell. This just converts it to `Hash`

File: chanrestore.go (hunk 10 above)
Comment (Crypt-iQ) [reply chanrestore.go]: Changed to use our `MultiSigKey`

File: tlv/complex.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package tlv
+
+import (
+	"io"
+
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// EKeyLocator is an encoder for keychain.KeyLocator.
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply tlv/complex.go:9]: That makes sense. There's a larger discussion I think that should be had about standalone libraries in the `lnd` repo because then they can't follow the mocking pattern in `lntest/mock` due to import cycles. Though that may be fixable by properly splitting up `lntest/mock`.

File: tlv/complex_test.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	r := bytes.NewReader(b.Bytes())
+	var decodedKeyLoc keychain.KeyLocator
+
+	if err := DKeyLocator(r, &decodedKeyLoc, &buf, 8); err != nil {
+		t.Fatalf("unable to decode %T: %v", &decodedKeyLoc, err)
+	}
+
+	// Finally, we'll compare that the original KeyLocator and the decoded
+	// version are equal.
+	if !reflect.DeepEqual(keyLoc, decodedKeyLoc) {
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply tlv/complex_test.go:44]: Yeah

**Comment by Crypt-iQ** [comment]
Looked into it a bit. Seems like threading a config flag to `initOurContribution` where the generation of `RevocationProducer` takes place will allow us to do full e2e tests with legacy SCBs.

File: tlv/complex.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply tlv/complex.go:9]: what sort of import cycles? i'm assuming you mean import cycles in test frameworks? this can be easily avoided by writing all tests for `<pkg>` in `<pkg>_test`.

an example of this is `wtdb`, where `wtmock` relies on types in `wtdb`. the test is in the same directory but uses the `wtdb_test` package, which a) is a work around for the import cycle, and b) ensures that the package's public api is tested, and isn't relying on internal members/functions/variables

[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply tlv/complex.go:9]: I guess it comes down to package-level mocking like `watchtower/wtmock` or the way I structured it in `lntest/mock`. `lntest/mock` will have more import cycles because everything isn't properly separated by package which is something I probably should have caught earlier.

[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply tlv/complex.go:9]: `channeldb` or `keychain`?

[the remark this answers is not in the document]
Comment (cfromknecht) [reply tlv/complex.go:9]: i'd do `channeldb` since that's where we keep all of our other serialization, and also the TLV type definition

File: lnwallet/wallet.go [hunk 13]
```diff
@@ -821,42 +821,14 @@ func (l *LightningWallet) initOurContribution(reservation *ChannelReservation,
...
-	// Perform an ECDH operation between the private key described in
-	// nextRevocationKeyDesc and our public multisig key. The result will be
-	// used to seed the revocation producer.
-	revRoot, err := l.ECDH(
-		nextRevocationKeyDesc, reservation.ourContribution.MultiSigKey.PubKey,
-	)
+	// With the above keys created, we'll also need to initialize our
+	// revocation tree state, and from that generate the
+	// per-commitment point.
+	producer, err := l.nextRevocationProducer(reservation, keyRing)
```
Comment (guggero) [inline lnwallet/wallet.go:827]: nit: extract this in the commit where we first change the code in this file? Or even in a separate commit before that?

**Review by guggero — COMMENTED** [review]
Code looks pretty good now, just a nit about commit structure.
Needs a rebase, going to do a full manual test run on this afterwards.

**Comment by Crypt-iQ** [comment]
@guggero Stale branch, must update

**Review by guggero — COMMENTED** [review]
This is getting very close, nice work!

I tested restoring both old and new channel backups and everything checks out.

File: chanrestore.go [hunk 14]
```diff
@@ -97,6 +84,47 @@ func (c *chanDBRestorer) openChannelShell(backup chanbackup.Single) (
...
+	// It currently has two possible formats.
+	var revRoot *chainhash.Hash
+
+	// If the PubKey field is non-nil, then this shachain root is using the
+	// legacy non-ECDH scheme.
+	if backup.ShaChainRootDesc.PubKey != nil {
+		ltndLog.Debugf("Using legacy revocation producer format for "+
+			"channel point %v", backup.FundingOutpoint)
+
+		// Obtain the private key for the shachain root from the encoded public
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanrestore.go:97]: line length is exactly 80

[the remark this answers is not in the document]
Comment (guggero) [reply chanrestore.go:97]: Not if you treat the tab character as 8 spaces: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#additional-style-constraints-on-top-of-gofmt
This confused me as well in the beginning.

[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanrestore.go]: Ah, gotcha

File: chanrestore.go [hunk 15]
```diff
@@ -97,6 +84,47 @@ func (c *chanDBRestorer) openChannelShell(backup chanbackup.Single) (
...
+		// key.
+		privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanrestore.go]: Updated log msg

**Review by guggero — APPROVED** [review]
Nice, LGTM :100:

**Comment by Roasbeef** [comment]
Needs a rebase!

**Comment by Crypt-iQ** [comment]
Rebased

File: lnwallet/revocation_producer_itest.go [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/shachain"
+)
+
+// nextRevocationProducer creates a new revocation producer, deriving the
+// revocation root by applying ECDH to a new key from our revocation root family
+// and the multisig key we use for the channel.
+func (l *LightningWallet) nextRevocationProducer(res *ChannelReservation,
```
Comment (Roasbeef) [inline lnwallet/revocation_producer_itest.go:14]: Very clever usage of build tags here!

**Review by Roasbeef — APPROVED** [review]
LGTM ☄️

File: lnwallet/revocation_producer_itest.go (hunk 16 above)
Comment (Crypt-iQ) [reply lnwallet/revocation_producer_itest.go:14]: All thanks to @guggero :D
